I've run this branch in a loop, like i mentioned in my pr.  like that one it fails about 25% of the time.

here are the errors i've seen so far:
```

gun3006.txt:java.lang.AssertionError: hasActiveWorkItems failed to see in time 3, had 1
gun3006.txt-	at me.itzg.tryetcdworkpart.services.WorkAllocatorTest.testTwoWorkersReleasedToOne(WorkAllocatorTest.java:280)
--
gun3007.txt:java.lang.AssertionError: hasActiveWorkItems failed to see in time 3, had 0
gun3007.txt-	at me.itzg.tryetcdworkpart.services.WorkAllocatorTest.testTwoWorkersReleasedToOne(WorkAllocatorTest.java:280)
--
gun3010.txt:java.lang.AssertionError: hasActiveWorkItems failed to see in time 3, had 0
gun3010.txt-	at me.itzg.tryetcdworkpart.services.WorkAllocatorTest.testTwoWorkersReleasedToOne(WorkAllocatorTest.java:280)
--
gun3012.txt:java.lang.AssertionError: hasActiveWorkItems failed to see in time 2, had 4
gun3012.txt-	at me.itzg.tryetcdworkpart.services.WorkAllocatorTest.testRebalanceRounding(WorkAllocatorTest.java:224)
--
gun3012.txt:java.lang.AssertionError: hasActiveWorkItems failed to see in time 3, had 0
gun3012.txt-	at me.itzg.tryetcdworkpart.services.WorkAllocatorTest.testTwoWorkersReleasedToOne(WorkAllocatorTest.java:281)
--
gun3013.txt:java.lang.AssertionError: hasActiveWorkItems failed to see in time 3, had 0
gun3013.txt-	at me.itzg.tryetcdworkpart.services.WorkAllocatorTest.testTwoWorkersReleasedToOne(WorkAllocatorTest.java:280)
```

--- a/src/main/java/me/itzg/tryetcdworkpart/services/WorkAllocator.java
+++ b/src/main/java/me/itzg/tryetcdworkpart/services/WorkAllocator.java
@@ -98,7 +98,7 @@ public class WorkAllocator implements SmartLifecycle {
   @Override
   public void start() {
     ourId = UUID.randomUUID().toString();
-    log.info("Staring WorkAllocator ourId={}", ourId);
+    log.info("Starting WorkAllocator ourId={}", ourId);
 
     running = true;
 

--- a/src/test/java/me/itzg/tryetcdworkpart/services/WorkAllocatorTest.java
+++ b/src/test/java/me/itzg/tryetcdworkpart/services/WorkAllocatorTest.java
@@ -48,6 +48,8 @@ import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
 @Slf4j
 public class WorkAllocatorTest {
 
+  private static final int TIMEOUT = 5000;
+
   @Rule
   public final EtcdClusterResource etcd = new EtcdClusterResource("test-etcd", 1);
 
@@ -109,7 +111,7 @@ public class WorkAllocatorTest {
     workAllocator.createWork("1")
     .join();
 
-    workProcessor.hasActiveWorkItems(1, 1000);
+    workProcessor.hasActiveWorkItems(1, TIMEOUT);
   }
 
   @Test
@@ -190,11 +192,11 @@ public class WorkAllocatorTest {
     }
 
     workAllocator1.start();
-    workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
+    workProcessor1.hasActiveWorkItems(totalWorkItems, TIMEOUT);
 
     workAllocator2.start();
-    workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
-    workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
+    workProcessor1.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
+    workProcessor2.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
 
   }
 
@@ -237,12 +239,12 @@ public class WorkAllocatorTest {
     }
 
     workAllocator1.start();
-    workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
+    workProcessor1.hasActiveWorkItems(totalWorkItems, TIMEOUT);
 
     log.info("starting second allocator");
     workAllocator2.start();
-    workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
-    workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
+    workProcessor1.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
+    workProcessor2.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
 
     log.info("starting third allocator");
     workAllocator3.start();
@@ -250,17 +252,17 @@ public class WorkAllocatorTest {
     Thread.sleep(600);
     // The expected balancing looks uneven here, but it's because the work load has a rounding-up
     // tolerance to avoid the one work item, in this case, from flip-flopping between the allocators
-    workProcessor1.hasActiveWorkItems(2, 1000);
-    workProcessor2.hasActiveWorkItems(2, 1000);
-    workProcessor3.hasActiveWorkItems(0, 1000);
+    workProcessor1.hasActiveWorkItems(2, TIMEOUT);
+    workProcessor2.hasActiveWorkItems(2, TIMEOUT);
+    workProcessor3.hasActiveWorkItems(0, TIMEOUT);
 
     log.info("adding new work for third to pickup");
     createdWork.add(
         workAllocator1.createWork(String.format("%d", totalWorkItems+1)).get()
     );
-    workProcessor3.hasActiveWorkItems(1, 1000);
-    workProcessor1.hasActiveWorkItems(2, 1000);
-    workProcessor2.hasActiveWorkItems(2, 1000);
+    workProcessor3.hasActiveWorkItems(1, TIMEOUT);
+    workProcessor1.hasActiveWorkItems(2, TIMEOUT);
+    workProcessor2.hasActiveWorkItems(2, TIMEOUT);
   }
 
   @Test
@@ -293,8 +295,8 @@ public class WorkAllocatorTest {
     }
 
     // first verify even load across the two
-    workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
-    workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
+    workProcessor1.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
+    workProcessor2.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
 
     log.info("stopping allocator 2");
     final Semaphore stopped = new Semaphore(0);
@@ -303,8 +305,8 @@ public class WorkAllocatorTest {
     });
 
     stopped.acquire();
-    workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
-    workProcessor2.hasActiveWorkItems(0, 1000);
+    workProcessor1.hasActiveWorkItems(totalWorkItems, TIMEOUT);
+    workProcessor2.hasActiveWorkItems(0, TIMEOUT);
   }
 
   private void assertWorkLoad(int expected, String workAllocatorId)

--- a/src/test/java/me/itzg/tryetcdworkpart/services/WorkAllocatorTest.java
+++ b/src/test/java/me/itzg/tryetcdworkpart/services/WorkAllocatorTest.java
@@ -194,6 +194,7 @@ public class WorkAllocatorTest {
     workAllocator1.start();
     workProcessor1.hasActiveWorkItems(totalWorkItems, TIMEOUT);
 
+    log.info("Starting second worker");
     workAllocator2.start();
     workProcessor1.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);
     workProcessor2.hasActiveWorkItems(totalWorkItems/2, TIMEOUT);

--- a/src/test/java/me/itzg/tryetcdworkpart/services/WorkAllocatorTest.java
+++ b/src/test/java/me/itzg/tryetcdworkpart/services/WorkAllocatorTest.java
@@ -58,6 +58,8 @@ public class WorkAllocatorTest {
   private WorkerProperties workerProperties;
   private Client client;
 
+  private List<WorkAllocator> workAllocatorsInTest;
+
   @Before
   public void setUp() throws Exception {
     taskExecutor = new ThreadPoolTaskScheduler();
@@ -72,19 +74,36 @@ public class WorkAllocatorTest {
         .map(URI::toString)
         .collect(Collectors.toList());
     client = Client.builder().endpoints(endpoints).build();
+
+    workAllocatorsInTest = new ArrayList<>();
   }
 
   @After
   public void tearDown() throws Exception {
-    taskExecutor.shutdown();
-    client.close();
+    log.info("Stopping work allocators");
+    final Semaphore stopSem = new Semaphore(0);
+
+    workAllocatorsInTest.forEach(workAllocator -> workAllocator.stop(() -> stopSem.release()));
+    try {
+      stopSem.tryAcquire(workAllocatorsInTest.size(), 5, TimeUnit.SECONDS);
+    } finally {
+      taskExecutor.shutdown();
+      client.close();
+    }
+  }
+
+  private WorkAllocator register(WorkAllocator workAllocator) {
+    workAllocatorsInTest.add(workAllocator);
+    return workAllocator;
   }
 
   @Test
   public void testSingleWorkItemSingleWorker() throws InterruptedException {
     final BulkWorkProcessor workProcessor = new BulkWorkProcessor(1);
-    final WorkAllocator workAllocator = new WorkAllocator(
-        workerProperties, client, workProcessor, taskExecutor);
+    final WorkAllocator workAllocator = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor, taskExecutor)
+    );
     workAllocator.start();
 
     workAllocator.createWork("1")
@@ -99,8 +118,10 @@ public class WorkAllocatorTest {
 
     final BulkWorkProcessor bulkWorkProcessor = new BulkWorkProcessor(totalWorkItems);
 
-    final WorkAllocator workAllocator = new WorkAllocator(
-        workerProperties, client, bulkWorkProcessor, taskExecutor);
+    final WorkAllocator workAllocator = register(
+        new WorkAllocator(
+        workerProperties, client, bulkWorkProcessor, taskExecutor)
+    );
 
     for (int i = 0; i < totalWorkItems; i++) {
       workAllocator.createWork(String.format("%d", i));
@@ -108,13 +129,9 @@ public class WorkAllocatorTest {
 
     workAllocator.start();
 
-    try {
-      bulkWorkProcessor.hasActiveWorkItems(totalWorkItems, 2000);
+    bulkWorkProcessor.hasActiveWorkItems(totalWorkItems, 2000);
 
-      assertWorkLoad(totalWorkItems, workAllocator.getId());
-    } finally {
-      workAllocator.stop();
-    }
+    assertWorkLoad(totalWorkItems, workAllocator.getId());
   }
 
   @Test
@@ -123,8 +140,10 @@ public class WorkAllocatorTest {
 
     final BulkWorkProcessor bulkWorkProcessor = new BulkWorkProcessor(totalWorkItems);
 
-    final WorkAllocator workAllocator = new WorkAllocator(
-        workerProperties, client, bulkWorkProcessor, taskExecutor);
+    final WorkAllocator workAllocator = register(
+        new WorkAllocator(
+        workerProperties, client, bulkWorkProcessor, taskExecutor)
+    );
 
     final List<Work> createdWork = new ArrayList<>();
     for (int i = 0; i < totalWorkItems; i++) {
@@ -135,17 +154,13 @@ public class WorkAllocatorTest {
 
     workAllocator.start();
 
-    try {
-      bulkWorkProcessor.hasActiveWorkItems(totalWorkItems, 2000);
-      assertWorkLoad(totalWorkItems, workAllocator.getId());
+    bulkWorkProcessor.hasActiveWorkItems(totalWorkItems, 2000);
+    assertWorkLoad(totalWorkItems, workAllocator.getId());
 
-      workAllocator.deleteWork(createdWork.get(0).getId());
-      bulkWorkProcessor.hasActiveWorkItems(totalWorkItems-1, 2000);
-      assertWorkLoad(totalWorkItems-1, workAllocator.getId());
+    workAllocator.deleteWork(createdWork.get(0).getId());
+    bulkWorkProcessor.hasActiveWorkItems(totalWorkItems-1, 2000);
+    assertWorkLoad(totalWorkItems-1, workAllocator.getId());
 
-    } finally {
-      workAllocator.stop();
-    }
   }
 
   @Test
@@ -156,33 +171,31 @@ public class WorkAllocatorTest {
     workerProperties.setRebalanceDelay(Duration.ofMillis(500));
 
     final BulkWorkProcessor workProcessor1 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator1 = new WorkAllocator(
-        workerProperties, client, workProcessor1, taskExecutor);
+    final WorkAllocator workAllocator1 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor1, taskExecutor)
+    );
 
     final BulkWorkProcessor workProcessor2 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator2 = new WorkAllocator(
-        workerProperties, client, workProcessor2, taskExecutor);
+    final WorkAllocator workAllocator2 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor2, taskExecutor)
+    );
 
-    try {
-      final List<Work> createdWork = new ArrayList<>();
-      for (int i = 0; i < totalWorkItems; i++) {
-        createdWork.add(
-            workAllocator1.createWork(String.format("%d", i)).get()
-        );
-      }
-
-      workAllocator1.start();
-      workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
-
-      workAllocator2.start();
-      workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
-      workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
-
+    final List<Work> createdWork = new ArrayList<>();
+    for (int i = 0; i < totalWorkItems; i++) {
+      createdWork.add(
+          workAllocator1.createWork(String.format("%d", i)).get()
+      );
     }
-    finally {
-      workAllocator1.stop();
-      workAllocator2.stop();
-    }
+
+    workAllocator1.start();
+    workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
+
+    workAllocator2.start();
+    workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
+    workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
+
   }
 
   @Test
@@ -193,61 +206,61 @@ public class WorkAllocatorTest {
     workerProperties.setRebalanceDelay(Duration.ofMillis(500));
 
     final BulkWorkProcessor workProcessor1 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator1 = new WorkAllocator(
-        workerProperties, client, workProcessor1, taskExecutor);
+    final WorkAllocator workAllocator1 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor1, taskExecutor)
+    );
 
     final BulkWorkProcessor workProcessor2 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator2 = new WorkAllocator(
-        workerProperties, client, workProcessor2, taskExecutor);
+    final WorkAllocator workAllocator2 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor2, taskExecutor)
+    );
 
     final BulkWorkProcessor workProcessor3 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator3 = new WorkAllocator(
-        workerProperties, client, workProcessor3, taskExecutor);
+    final WorkAllocator workAllocator3 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor3, taskExecutor)
+    );
 
     final BulkWorkProcessor workProcessor4 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator4 = new WorkAllocator(
-        workerProperties, client, workProcessor4, taskExecutor);
+    final WorkAllocator workAllocator4 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor4, taskExecutor)
+    );
 
-    try {
-      final List<Work> createdWork = new ArrayList<>();
-      for (int i = 0; i < totalWorkItems; i++) {
-        createdWork.add(
-            workAllocator1.createWork(String.format("%d", i)).get()
-        );
-      }
-
-      workAllocator1.start();
-      workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
-
-      log.info("starting second allocator");
-      workAllocator2.start();
-      workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
-      workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
-
-      log.info("starting third allocator");
-      workAllocator3.start();
-      // wait past rebalance to ensure load balance of 1.3333 is rounded to 2
-      Thread.sleep(600);
-      // The expected balancing looks uneven here, but it's because the work load has a rounding-up
-      // tolerance to avoid the one work item, in this case, from flip-flopping between the allocators
-      workProcessor1.hasActiveWorkItems(2, 1000);
-      workProcessor2.hasActiveWorkItems(2, 1000);
-      workProcessor3.hasActiveWorkItems(0, 1000);
-
-      log.info("adding new work for third to pickup");
+    final List<Work> createdWork = new ArrayList<>();
+    for (int i = 0; i < totalWorkItems; i++) {
       createdWork.add(
-          workAllocator1.createWork(String.format("%d", totalWorkItems+1)).get()
+          workAllocator1.createWork(String.format("%d", i)).get()
       );
-      workProcessor3.hasActiveWorkItems(1, 1000);
-      workProcessor1.hasActiveWorkItems(2, 1000);
-      workProcessor2.hasActiveWorkItems(2, 1000);
     }
-    finally {
-      log.info("stopping allocators");
-      workAllocator1.stop();
-      workAllocator2.stop();
-      workAllocator3.stop();
-    }
+
+    workAllocator1.start();
+    workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
+
+    log.info("starting second allocator");
+    workAllocator2.start();
+    workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
+    workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
+
+    log.info("starting third allocator");
+    workAllocator3.start();
+    // wait past rebalance to ensure load balance of 1.3333 is rounded to 2
+    Thread.sleep(600);
+    // The expected balancing looks uneven here, but it's because the work load has a rounding-up
+    // tolerance to avoid the one work item, in this case, from flip-flopping between the allocators
+    workProcessor1.hasActiveWorkItems(2, 1000);
+    workProcessor2.hasActiveWorkItems(2, 1000);
+    workProcessor3.hasActiveWorkItems(0, 1000);
+
+    log.info("adding new work for third to pickup");
+    createdWork.add(
+        workAllocator1.createWork(String.format("%d", totalWorkItems+1)).get()
+    );
+    workProcessor3.hasActiveWorkItems(1, 1000);
+    workProcessor1.hasActiveWorkItems(2, 1000);
+    workProcessor2.hasActiveWorkItems(2, 1000);
   }
 
   @Test
@@ -257,43 +270,41 @@ public class WorkAllocatorTest {
     workerProperties.setRebalanceDelay(Duration.ofMillis(500));
 
     final BulkWorkProcessor workProcessor1 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator1 = new WorkAllocator(
-        workerProperties, client, workProcessor1, taskExecutor);
+    final WorkAllocator workAllocator1 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor1, taskExecutor)
+    );
     workAllocator1.start();
 
     final BulkWorkProcessor workProcessor2 = new BulkWorkProcessor(totalWorkItems);
-    final WorkAllocator workAllocator2 = new WorkAllocator(
-        workerProperties, client, workProcessor2, taskExecutor);
+    final WorkAllocator workAllocator2 = register(
+        new WorkAllocator(
+        workerProperties, client, workProcessor2, taskExecutor)
+    );
     workAllocator2.start();
 
     // both are started and ready to accept newly created worked items
 
-    try {
-      final List<Work> createdWork = new ArrayList<>();
-      for (int i = 0; i < totalWorkItems; i++) {
-        createdWork.add(
-            workAllocator1.createWork(String.format("%d", i)).get()
-        );
-      }
-
-      // first verify even load across the two
-      workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
-      workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
-
-      log.info("stopping allocator 2");
-      final Semaphore stopped = new Semaphore(0);
-      workAllocator2.stop(() -> {
-        stopped.release();
-      });
-
-      stopped.acquire();
-      workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
-      workProcessor2.hasActiveWorkItems(0, 1000);
+    final List<Work> createdWork = new ArrayList<>();
+    for (int i = 0; i < totalWorkItems; i++) {
+      createdWork.add(
+          workAllocator1.createWork(String.format("%d", i)).get()
+      );
     }
-    finally {
-      workAllocator1.stop();
-      workAllocator2.stop();
-    }
+
+    // first verify even load across the two
+    workProcessor1.hasActiveWorkItems(totalWorkItems/2, 1000);
+    workProcessor2.hasActiveWorkItems(totalWorkItems/2, 1000);
+
+    log.info("stopping allocator 2");
+    final Semaphore stopped = new Semaphore(0);
+    workAllocator2.stop(() -> {
+      stopped.release();
+    });
+
+    stopped.acquire();
+    workProcessor1.hasActiveWorkItems(totalWorkItems, 1000);
+    workProcessor2.hasActiveWorkItems(0, 1000);
   }
 
   private void assertWorkLoad(int expected, String workAllocatorId)
@@ -321,7 +332,7 @@ public class WorkAllocatorTest {
     final BulkWorkProcessor bulkWorkProcessor = new BulkWorkProcessor(totalWorkItems);
 
     final List<WorkAllocator> workAllocators = IntStream.range(0, totalWorkers)
-        .mapToObj(index -> new WorkAllocator(workerProperties, client, bulkWorkProcessor, taskExecutor))
+        .mapToObj(index -> register(new WorkAllocator(workerProperties, client, bulkWorkProcessor, taskExecutor)))
         .collect(Collectors.toList());
 
     for (WorkAllocator workAllocator : workAllocators) {
@@ -367,11 +378,6 @@ public class WorkAllocatorTest {
     workItemSummary.workerLoad.forEach((workerId, load) -> {
       log.info("Worker={} has load={}", workerId, load);
     });
-
-    Thread.sleep(1000);
-    for (WorkAllocator workAllocator : workAllocators) {
-      workAllocator.stop();
-    }
   }
 
   @Test

